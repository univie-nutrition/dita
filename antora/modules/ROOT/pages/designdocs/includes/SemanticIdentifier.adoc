= Semantic Identifier

A _Semantic Identifier_ (SID) references data objects across system boundaries.

[plantuml,fig-sid,svg]
.Semantic Identifier (SID)
----
@startuml

class SystemId {
    system: String
    version: String
}

class ObjectId {
    context: String
    objectSimpleId: String
}

class SemanticIdentifier {
}

class SemanticIdentifierSet {
}

note right of SemanticIdentifier::systemId
  identifies the system (and optionally system version) of the referenced data object
end note
note right of SemanticIdentifier::objectId
  uniquely identifies the data object within the system
end note

SemanticIdentifier --> SystemId
SemanticIdentifier --> ObjectId

SemanticIdentifierSet -right-> "*" SemanticIdentifier : has many

hide class method

@enduml
----

Example use, referencing a food, product or facet.

[source,java]
.Java
----
// within BLS v3.02 represents 'Banana raw'
var blsBanana = SemanticIdentifier.parse("de.bls/3.02", "food/F503100");
var totalSuguar = SemanticIdentifier.parse("de.bls/3.02", "comp/KMD");

// within LanguaL 2017 Facet Category - A (PRODUCT TYPE) represents 'Bananas' from GS1 ontology
var gs1BananaId = new SemanticIdentifier("langual/2017:A1415"); 
----