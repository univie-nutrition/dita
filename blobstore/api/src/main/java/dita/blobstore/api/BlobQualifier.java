/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.blobstore.api;

import org.apache.causeway.commons.collections.Can;
import org.apache.causeway.commons.internal.assertions._Assert;

/**
 * Has non-empty name, with characters conforming to {@link Character#isJavaIdentifierPart(int)}.
 * <p>
 * A blob may provide zero ore more qualifiers via its descriptor.
 * A blob-store search can then be refined to filter for results that satisfy <b>all</b> the requested
 * qualifiers.
 */
public record BlobQualifier(
        String name) {

    public static Can<BlobQualifier> of(final String ...names) {
        return Can.ofArray(names).map(BlobQualifier::new);
    }

    public BlobQualifier(final String name) {
        _Assert.assertNotEmpty(name);
        _Assert.assertTrue(name.chars().allMatch(Character::isJavaIdentifierPart));
        this.name = name;
    }
}
