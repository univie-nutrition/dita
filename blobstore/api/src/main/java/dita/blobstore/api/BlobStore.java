/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.blobstore.api;

import java.util.Optional;

import org.springframework.lang.Nullable;

import org.apache.causeway.applib.value.Blob;
import org.apache.causeway.commons.collections.Can;

import dita.commons.types.NamedPath;
import lombok.NonNull;

public interface BlobStore {

    void putBlob(@NonNull BlobDescriptor blobDescriptor, @NonNull Blob blob);

    Can<BlobDescriptor> listDescriptors(@Nullable NamedPath path, boolean recursive);
    Optional<BlobDescriptor> lookupDescriptor(@Nullable NamedPath path);
    Optional<Blob> lookupBlob(@Nullable NamedPath path);
    void deleteBlob(@Nullable NamedPath path);

}
