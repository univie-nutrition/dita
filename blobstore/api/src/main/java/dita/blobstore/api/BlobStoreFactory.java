/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.blobstore.api;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.AutowireCapableBeanFactory;
import org.springframework.stereotype.Service;

import lombok.SneakyThrows;

@Service
public class BlobStoreFactory {

    public static record BlobStoreConfiguration(
            Class<? extends BlobStore> implementation,
            String resource) {
    }

    private @Autowired AutowireCapableBeanFactory beanFactory;

    @SneakyThrows
    public BlobStore createBlobStore(final BlobStoreConfiguration config) {
        var blobStore = config.implementation().getConstructor(new Class[]{BlobStoreConfiguration.class})
            .newInstance(config);
        beanFactory.autowireBean(blobStore);
        return blobStore;
    }

}
