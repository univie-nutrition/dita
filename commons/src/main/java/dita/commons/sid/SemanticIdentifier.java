/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.commons.sid;

import java.util.function.UnaryOperator;

import org.springframework.lang.Nullable;

import org.apache.causeway.commons.internal.base._Strings;

import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.experimental.Accessors;

/**
 * A Semantic Identifier references data objects across system boundaries.
 * <p>
 * <ul>
 * <li>system: at.gd</li>
 * <li>version: 2.0</li>
 * <li>context: food | recipe | food.desc | recipe.desc | food.group etc.</li>
 * <li>object: 00123</li>
 * </ul>
 * <p>
 * formats to -> SID[at.gd/2.0, food/00123]
 */
@Builder
public record SemanticIdentifier (
        /**
         * Identifies the system of the referenced data object.
         * (optionally includes version information)
         */
        @NonNull SystemId systemId,
        /**
         * Uniquely identifies the data object within the system.
         * (optionally includes context information)
         */
        @NonNull ObjectId objectId) implements Comparable<SemanticIdentifier> {

    // -- SYSTEM ID

    public record SystemId(
            /**
             * Identifies the system of the referenced data object.
             */
            String system,
            /**
             * Identifies the system's version of the referenced data object.
             */
            String version) implements Comparable<SystemId> {

        public static SystemId empty() {
            return new SystemId(null, null);
        }

        public static SystemId parse(final @Nullable String systemIdStringified) {
            return _Utils.parseSystemId(systemIdStringified);
        }

        public SystemId(final String system, final String version) {
            this.system = _Utils.validate(system);
            this.version = _Utils.validate(version);
        }

        /**
         * SystemId constructor with empty version part.
         */
        public SystemId(final String system) {
            this(system, null);
        }

        // -- WITHER

        public SystemId withSystem(final String system) {
            return new SystemId(system, version());
        }
        public SystemId withVersion(final String version) {
            return new SystemId(system(), version);
        }

        // -- MAPPER

        public SystemId mapSystem(final UnaryOperator<String> systemMapper) {
            return withSystem(systemMapper.apply(system()));
        }
        public SystemId mapVersion(final UnaryOperator<String> versionMapper) {
            return withVersion(versionMapper.apply(version()));
        }

        // -- CONTRACT

        @Override
        public int compareTo(final @Nullable SystemId o) {
            return compare(this, o);
        }

        /**
         * Identifies the system of the referenced data object.
         * (optionally suffixed with a version)
         */
        @Override
        public final String toString() {
            return _Utils.format(this);
        }

        public static int compare(
                final @Nullable SystemId a,
                final @Nullable SystemId b) {
            if(a==null) return b==null
                        ? 0
                        : -1;
            if(b==null) return 1;
            final int c = _Strings.compareNullsFirst(a.system(), b.system());
            return c!=0
                ? c
                // TODO should use DEWEY compare here
                : _Strings.compareNullsFirst(a.version(), b.version());
        }

    }

    // -- OBJECT ID

    public record ObjectId(
            /**
             * Uniquely identifies the data object's context within the system.
             */
            String context,
            /**
             * Uniquely identifies the data object within the system's context.
             */
            String object) implements Comparable<ObjectId> {

        /**
         * Some predefined contexts.
         */
        @RequiredArgsConstructor
        public enum Context {
            LITERAL("lit"),
            LANGUAGE("lang"),
            COMPONENT("comp"),
            FOOD("food"),
            RECIPE("recp"),
            BRAND("brand"),
            FOOD_DESCRIPTOR("fd"),
            RECIPE_DESCRIPTOR("rd"),
            FOOD_GROUP("fg"),
            RECIPE_GROUP("rg"),
            ;
            @Getter @Accessors(fluent=true)
            final String id;
        }

        public static ObjectId empty() {
            return new ObjectId((String)null, null);
        }

        public static ObjectId parse(final @Nullable String objectIdStringified) {
            return _Utils.parseObjectId(objectIdStringified);
        }

        public ObjectId(final String context, final String object) {
            this.context = _Utils.validate(context);
            this.object = _Utils.validateObject(object);
        }

        public ObjectId(final @NonNull Context context, final String object) {
            this(context.id, object);
        }

        /**
         * ObjectId constructor with empty context part.
         */
        public ObjectId(final String object) {
            this((String)null, object);
        }

        // -- WITHER

        public ObjectId withContext(final String context) {
            return new ObjectId(context, object());
        }
        public ObjectId withObject(final String object) {
            return new ObjectId(context(), object);
        }

        // -- MAPPER

        public ObjectId mapContext(final UnaryOperator<String> contextMapper) {
            return withContext(contextMapper.apply(context()));
        }
        public ObjectId mapObject(final UnaryOperator<String> objectMapper) {
            return withObject(objectMapper.apply(object()));
        }

        // -- CONTRACT

        @Override
        public int compareTo(final @Nullable ObjectId o) {
            return compare(this, o);
        }

        /**
         * Uniquely identifies the data object within the system.
         * (optionally prefixed with a context)
         */
        @Override
        public final String toString() {
            return _Utils.format(this);
        }

        public static int compare(
                final @Nullable ObjectId a,
                final @Nullable ObjectId b) {
            if(a==null) return b==null
                        ? 0
                        : -1;
            if(b==null) return 1;
            final int c = _Strings.compareNullsFirst(a.context(), b.context());
            return c!=0
                ? c
                // TODO should use DEWEY compare here
                : _Strings.compareNullsFirst(a.object(), b.object());
        }

    }

    // -- SID IMPL

    public static SemanticIdentifier empty() {
        return new SemanticIdentifier(SystemId.empty(), ObjectId.empty());
    }

    public SemanticIdentifier(final String systemId, final String objectId) {
        this(SystemId.parse(systemId), ObjectId.parse(objectId));
    }

    // -- PARSER

    /**
     * Inverse of {@link SemanticIdentifier#toString()}.
     * @see SemanticIdentifier#toString()
     */
    public static SemanticIdentifier parse(final @Nullable String stringified) {
        return _Utils.parseSid(stringified);
    }

    // -- WITHER

    public SemanticIdentifier withSystemId(final SystemId systemId) {
        return new SemanticIdentifier(systemId, objectId());
    }
    public SemanticIdentifier withObjectId(final ObjectId objectId) {
        return new SemanticIdentifier(systemId(), objectId);
    }

    // -- MAPPER

    public SemanticIdentifier mapSystemId(final UnaryOperator<SystemId> systemIdMapper) {
        return withSystemId(systemIdMapper.apply(systemId()));
    }
    public SemanticIdentifier mapObjectId(final UnaryOperator<ObjectId> objectIdMapper) {
        return withObjectId(objectIdMapper.apply(objectId()));
    }


    @Override
    public int compareTo(final @Nullable SemanticIdentifier o) {
        return compare(this, o);
    }

    /**
     * e.g. {@code at.gd/2.0:food/00123}
     */
    public String toStringNoBox() {
        return _Utils.formatUnboxed(this);
    }

    /**
     * e.g. {@code SID[at.gd/2.0:food/00123]}
     */
    @Override
    public final String toString() {
        return _Utils.formatBoxed(this);
    }

    // -- UTILITY

    public static int compare(
            final @Nullable SemanticIdentifier a,
            final @Nullable SemanticIdentifier b) {
        if(a==null) return b==null
                    ? 0
                    : -1;
        if(b==null) return 1;
        final int c = a.systemId().compareTo(b.systemId());
        return c!=0
            ? c
            : a.objectId().compareTo(b.objectId());
    }

}
