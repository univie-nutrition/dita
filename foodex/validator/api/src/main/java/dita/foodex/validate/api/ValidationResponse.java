/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.foodex.validate.api;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import org.apache.causeway.commons.internal.base._NullSafe;
import org.apache.causeway.commons.internal.base._Strings;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data @AllArgsConstructor(staticName = "of") @NoArgsConstructor
public class ValidationResponse {
    
    @Data @AllArgsConstructor(staticName = "of") @NoArgsConstructor
    public static class Message {
        String message;
        String risk;
        String risk2;
        
        public static Message failure(String msg) {
            return Message.of(msg, "ERROR", "ERROR");
        }
        
    }
    
    String fullCode;
    
    List<Message> messages;
    
    boolean isEmpty;
    
    public String asSingleMessage() {
        return _NullSafe.stream(getMessages())
        .map(msg->String.format("%s ;%s;%s", msg.getMessage(), msg.getRisk(), msg.getRisk2()))
        .collect(Collectors.joining("|"));
    }
    
    public static ValidationResponse parse(String rawValidation, String fullCode) {
        if(_NullSafe.isEmpty(rawValidation)) {
            return empty();
        }
        
        var messages = _Strings.splitThenStream(rawValidation, "|")
                .map(String::trim)
                .filter(_Strings::isNotEmpty)
                .<Message>map(rawMessage->{
            
                    var chunks = _Strings.splitThenStream(rawMessage, ";")
                            .map(String::trim)
                            .collect(Collectors.toList());
                    
                    if(chunks.size()!=3) {
                        return Message.of(rawMessage, "UNKNOWN", "UNKNOWN");
                    }
                    
                    return Message.of(chunks.get(0), chunks.get(1), chunks.get(2));
                    
                })
                .collect(Collectors.toList());
        
        if(messages.isEmpty()) {
            return empty();
        }
        
        return ValidationResponse.of(fullCode, messages, false);
    }
    
    public static ValidationResponse empty() {
        return ValidationResponse.of(null, null, true);
    }

    public static ValidationResponse failure(String fullCode, Exception failureCause) {
        return ValidationResponse.of(
                fullCode, 
                Collections.singletonList(Message.failure(failureCause.getMessage())), false);
    }
    
}
