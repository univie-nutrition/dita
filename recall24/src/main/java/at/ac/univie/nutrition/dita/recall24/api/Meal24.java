/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package at.ac.univie.nutrition.dita.recall24.api;

import java.time.LocalTime;

import org.apache.causeway.commons.collections.Can;

// TODO: Auto-generated Javadoc
/**
 * The Interface Meal24.
 */
public interface Meal24 {

    /**
     * Parent interview.
     */
    Interview24 parentInterview();

    /**
     * Hour of day, when this meal took place.
     */
    LocalTime hourOfDay();

    /**
     * Identifying the occasion, when this meal took place.
     */
    String foodConsumptionOccasionId();

    /**
     * Identifying the place, where this meal took place.
     */
    String foodConsumptionPlaceId();

    /**
     * Memorized food for this meal.
     */
    Can<? extends MemorizedFood24> memorizedFood();

}