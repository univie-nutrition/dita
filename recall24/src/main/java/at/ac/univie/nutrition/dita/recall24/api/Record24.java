/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package at.ac.univie.nutrition.dita.recall24.api;

import org.apache.causeway.commons.collections.Can;

public interface Record24 {

    //TODO - it should be enough to distinguish between just 2: REGULAR and INFORMAL
    public static enum Type {

        /**
         * Also used as Proxy to 'Composite Food' (eg. gd-recipe-manager) or 'Product'
         */
        FOOD,

        /**
         * Recipe with ingredients, where the recipe itself and its ingredients have facets.
         */
        COMPOSITE,

        PRODUCT,

        INFORMAL
        ;

        public boolean isFoodOrProxy() { return this == FOOD; }
        public boolean isComposite() { return this == COMPOSITE; }
        public boolean isProduct() { return this == PRODUCT; }
        public boolean isInformal() { return this == INFORMAL; }
    }

    /**
     * Memorized food this record belongs to.
     */
    MemorizedFood24 parentMemorizedFood();

    /**
     * The type of this record.
     */
    Record24.Type type();

    /**
     * The name of this record.
     */
    String name();

    /**
     * Comma separated list of facet identifiers,
     * ordered (by some natural order).
     */
    String facetSids();

    /**
     * Ingredients of this record.
     */
    Can<? extends Ingredient24> ingredients();

}