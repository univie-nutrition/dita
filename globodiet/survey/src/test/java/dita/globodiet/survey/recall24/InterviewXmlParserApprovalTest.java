/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.recall24;

import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import dita.commons.sid.SemanticIdentifier.SystemId;
import dita.globodiet.survey.utils.ApprovalTestOptions;

class InterviewXmlParserApprovalTest {

    private SystemId systemId = new SystemId("at.gd", "2.0");

    @Test
    @UseReporter(DiffReporter.class)
    void parsingSample() {
        var interviewSet24 = InterviewSamples.SAMPLES.asInterviewSet(systemId, null);
        Approvals.verify(interviewSet24.toYaml(), ApprovalTestOptions.yamlOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void parsingRefComposites() {
        var interviewSet24 = InterviewSamples.COMPOSITES.asInterviewSet(systemId, null);
        Approvals.verify(interviewSet24.toYaml(), ApprovalTestOptions.yamlOptions());
    }

    @Test
    @UseReporter(DiffReporter.class)
    void parsingRefFatSouceSweetener() {
        var interviewSet24 = InterviewSamples.FAT_SOUCE_SWEETENERS.asInterviewSet(systemId, null);
        Approvals.verify(interviewSet24.toYaml(), ApprovalTestOptions.yamlOptions());
    }

}
