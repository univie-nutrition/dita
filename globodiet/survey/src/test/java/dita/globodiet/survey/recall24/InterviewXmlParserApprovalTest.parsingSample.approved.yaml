respondents:
- alias: "01_01_SL_2"
  dateOfBirth: "1933-03-17"
  sex: "FEMALE"
  interviews:
  - interviewDate: "2021-05-02"
    consumptionDate: "2021-04-06"
    interviewOrdinalRef:
      value: 1
    respondentSupplementaryData:
      specialDietId: "99"
      specialDayId: "00"
      heightCM: 160
      weightKG: 54.9
    meals:
    - hourOfDay: "07:00:00"
      foodConsumptionOccasionId: "01"
      foodConsumptionPlaceId: "01"
      memorizedFood:
      - name: "Kaffee"
        topLevelRecords:
        - type: "COMPOSITE"
          name: "Eieraufstrich"
          sid: "at.gd/2.0:recp/00232"
          facetSids: "at.gd/2.0:rd/0103, at.gd/2.0:rd/0200, at.gd/2.0:rd/0317, at.gd/2.0:rd/0405"
          subRecords:
          - type: "FOOD"
            name: "Mayonnaise"
            sid: "at.gd/2.0:food/01529"
            facetSids: "at.gd/2.0:fd/0400, at.gd/2.0:fd/0699, at.gd/2.0:fd/0801, at.gd/2.0:fd/1000"
            amountConsumed: 6.7415728569030762
            consumptionUnit: "GRAM"
            rawPerCookedRatio: 1
            typeOfFatUsedDuringCooking: null
            typeOfMilkOrLiquidUsedDuringCooking: null
            annotations:
              group: "at.gd/2.0:fg/150102"
          - type: "FOOD"
            name: "Ei"
            sid: "at.gd/2.0:food/00916"
            facetSids: "at.gd/2.0:fd/0107, at.gd/2.0:fd/0303"
            amountConsumed: 16.853933334350586
            consumptionUnit: "GRAM"
            rawPerCookedRatio: 1
            typeOfFatUsedDuringCooking: null
            typeOfMilkOrLiquidUsedDuringCooking: null
            annotations:
              group: "at.gd/2.0:fg/0901"
          - type: "FOOD"
            name: "Schnittlauch"
            sid: "at.gd/2.0:food/01547"
            facetSids: "at.gd/2.0:fd/0205, at.gd/2.0:fd/0499"
            amountConsumed: 0.93632960319519043
            consumptionUnit: "GRAM"
            rawPerCookedRatio: 1
            typeOfFatUsedDuringCooking: null
            typeOfMilkOrLiquidUsedDuringCooking: null
            annotations:
              group: "at.gd/2.0:fg/1503"
          - type: "FOOD"
            name: "Knoblauch"
            sid: "at.gd/2.0:food/00103"
            facetSids: "at.gd/2.0:fd/0205, at.gd/2.0:fd/0399, at.gd/2.0:fd/0499, at.gd/2.0:fd/0999,\
              \ at.gd/2.0:fd/1200, at.gd/2.0:fd/2800"
            amountConsumed: 0.37453183531761169
            consumptionUnit: "GRAM"
            rawPerCookedRatio: 1
            typeOfFatUsedDuringCooking: null
            typeOfMilkOrLiquidUsedDuringCooking: null
            annotations:
              group: "at.gd/2.0:fg/0207"
          - type: "FOOD"
            name: "Salz, Koch"
            sid: "at.gd/2.0:food/01571"
            facetSids: ""
            amountConsumed: 0.093632958829402924
            consumptionUnit: "GRAM"
            rawPerCookedRatio: 1
            typeOfFatUsedDuringCooking: null
            typeOfMilkOrLiquidUsedDuringCooking: null
            annotations:
              group: "at.gd/2.0:fg/1504"
          annotations:
            group: "at.gd/2.0:rg/03"
        - type: "FOOD"
          name: "Kaffee, mit Koffein"
          sid: "at.gd/2.0:food/01339"
          facetSids: "at.gd/2.0:fd/0201, at.gd/2.0:fd/0699, at.gd/2.0:fd/0702, at.gd/2.0:fd/1200"
          amountConsumed: 125
          consumptionUnit: "GRAM"
          rawPerCookedRatio: 1
          typeOfFatUsedDuringCooking: null
          typeOfMilkOrLiquidUsedDuringCooking: null
          annotations:
            group: "at.gd/2.0:fg/130301"
    annotations:
      dataSource: "Byte-Resource[3c 3f 78 6d 6c 20 76 65 72 73 69 6f 6e 3d 22 31 ...]"
annotations: {}
